### PR TITLE
nrf9151 preview

### PR DESCRIPTION
**Pull request by nerwalt**
Adding support for the nrf9120 (nrf9151 and nrf9161)

### CODE REVIEW

**Comment by nerwalt** [comment]
This is a draft/WIP PR.  It depends on nrflixb-sys getting updated to v2.7.0.  I submitted a PR for that earlier.  Also, a few new FFI functions need to be filled in

File: src/lib.rs [hunk 1]
```diff
@@ -125,28 +132,29 @@ pub async fn init_with_custom_layout(
...
             trace: nrfxlib_sys::nrf_modem_shmem_cfg__bindgen_ty_4 {
                 base: memory_layout.base_address
-                    + nrfxlib_sys::NRF_MODEM_SHMEM_CTRL_SIZE
+                    + nrfxlib_sys::NRF_MODEM_CELLULAR_SHMEM_CTRL_SIZE
                     + memory_layout.tx_area_size
                     + memory_layout.rx_area_size,
                 size: memory_layout.trace_area_size,
             },
         },
-        ipc_irq_prio: 0,
```
Comment (diondokter) [inline src/lib.rs:148]: Why this change?

Comment (nerwalt) [reply src/lib.rs:148]: I made that change when I started working with the nrf9151, and I don't remember why.  It isn't necessary, so I will revert it.

File: src/ffi.rs [hunk 2]
```diff
@@ -524,39 +524,153 @@ pub extern "C" fn nrf_modem_os_is_in_isr() -> bool {
...
+        *mutex = nrf_modem_os_alloc(core::mem::size_of::<MutexLock>()) as *mut _;
+
+        if (*mutex).is_null() {
+            // We are out of memory
+            return -(nrfxlib_sys::NRF_ENOMEM as i32);
+        }
+    } 
+
+    // (Re-)Initialize (unlock) the mutex
+    (*(mutex as *mut MutexLock)).unlock();
```
Comment (diondokter) [inline src/ffi.rs:576]: Technically speaking the `MutexLock` is uninitialized.
I think it'd be better to put it in a `MaybeUninit` and write a fresh properly initialized `MutexLock` into it.

Does it really need to? Well... maybe? I don't know if `AtomicBool` is fine with random data. I think it's UB for a normal bool.
But even if it is fine, using `MaybeUninit` is more proper.

**Review by diondokter — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -8,25 +8,32 @@ description = "Async modem api for the nRF9160"
...
 repository = "https://github.com/diondokter/nrf-modem"
 readme = "README.md"
-keywords = ["nRF9160", "LTE", "GPS", "NB-IoT", "embedded"]
+keywords = ["nRF9160", "nRF9151", "LTE", "GPS", "NB-IoT", "embedded"]
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-nrfxlib-sys = "2.4.2"
+nrfxlib-sys = "2.7.0"
```
Comment (diondokter) [inline Cargo.toml:16]: Please make this pinned. I should have done that with the previous version already.
```suggestion
nrfxlib-sys = "=2.7.0"
```

**Comment by diondokter** [comment]
Can we think of a reason why this would be a breaking change? IIRC we don't forward any types from `nrfxlib-sys`...
But maybe just bump it to make sure and make this version 0.5?

File: src/ffi.rs (hunk 2 above)
Comment (nerwalt) [reply src/ffi.rs:576]: That is right.  Updated.

**Comment by nerwalt** [comment]
I am unsure what to do about `nrf_modem_os_log` and `nrf_modem_os_logdump`.  I don't need them right now and they aren't required, so leaving them as stubs seems fine.  I have seen a few other crates provide logging hooks with a thin C function.  For example, https://github.com/Ragarnoy/a121-sys.  The same could be done here and hook it into defmt.

**Comment by diondokter** [comment]
> I am unsure what to do about `nrf_modem_os_log` and `nrf_modem_os_logdump`. I don't need them right now and they aren't required, so leaving them as stubs seems fine. I have seen a few other crates provide logging hooks with a thin C function (to handle the variadic args). For example, https://github.com/Ragarnoy/a121-sys. The same could be done here and hook it into defmt.

Yeah just leave them as stubs for now. That's fine.
If someone wants to implement it, then they can send a PR.

**Review by diondokter — APPROVED** [review]
